src/main/combatLogParser.ts

```diff
@@ -0,0 +1,434 @@
+    private _handlerLockTimeout: any;
+
+    private readonly _tailOptions = {
+        flushAtEOF: true
+    };
+
+    constructor(options: CombatLogParserOptionsType) {
+        super();
+
+        this._options = options;
+    };
+
+    /**
+     * Start watching a path, if it's a valid WoW logs directory.
+     * This is checked via `getWowFlavour()` which looks for the file '../.flavour.info'
+     * relative to the log directory, which all flavours of WoW have.
+     */
+    watchPath(pathSpec: string): void {
+        pathSpec = path.resolve(pathSpec);
+        if (pathSpec in this._handlers) {
+            this.unwatchPath(pathSpec);
+        }
+
+        const wowFlavour = this.getWowFlavour(pathSpec);
+        if (wowFlavour === 'unknown') {
+            console.warn(`[CombatLogParser] Ignoring non-WoW combat log directory '${pathSpec}'`);
+            return;
+        }
+
+        this._handlers[pathSpec] = {
+            wowFlavour,
+            path: pathSpec,
+        };
+
+        console.log(`[CombatLogParser] Start watching '${pathSpec}' for '${wowFlavour}'`, this._handlers[pathSpec]);
+        this.watchLogDirectory(pathSpec);
+    }
+
+    /**
+     * Unwatch a previously watched path.
+     */
+    unwatchPath(pathSpec: string): void {
+        pathSpec = path.resolve(pathSpec);
+        if (!(pathSpec in this._handlers)) {
+            return;
+        }
+
+        clearInterval(this._handlers[pathSpec].watchInterval);
+
+        if (this._handlers[pathSpec].tail) {
+            this._handlers[pathSpec].tail.unwatch();
+        }
+
+        delete this._handlers[pathSpec];
+
+        console.log(`[CombatLogParser] Stop watching '${pathSpec}'`);
+    }
+
+    /**
+     * Unwatch all paths.
+     */
+    unwatch(): void {
+        Object.keys(this._handlers).forEach(this.unwatchPath.bind(this));
+    }
+
+    /**
+     * Handle a line from the WoW log.
+     */
+     handleLogLine(flavour: string, line: string) {
+        const logLine = new LogLine(line)
+        const logEventType = logLine.type();
+
+        this.emit(logEventType, logLine, flavour);
+    }
+
+    /**
+     * Find and return the flavour of WoW that the log directory
+     * belongs to by means of the '.flavor.info' file.
+     */
+     private getWowFlavour(pathSpec: string): string {
+        const flavourInfoFile = path.normalize(
+            path.join(pathSpec, '../.flavor.info')
+        );
+
+        // If this file doesn't exist, it's not a WoW combat log directory
+        if (!fs.existsSync(flavourInfoFile)) {
+            return 'unknown';
+        }
+
+        const content = fs.readFileSync(flavourInfoFile)
+            .toString()
+            .split("\n");
+
+        return content.length > 1 ? content[1] : 'unknown';
+    }
+
+    /**
+     * Ensure only a single logfile is being watched once one of them starts
+     * receiving data.
+     *
+     * The lock will timeout after 30 seconds if no data has been received.
```

Comment:
Interesting. I think I was expecting the lock to be released by some other timeout. Might discover this as I read on, but are we stopping the recording if a lock times out? We probably should be. Is 30s a reasonable time for that? We might need some real data here to be sure. I could imagine a scenario where some M+ group takes a toilet break (they are casuals for arguments sake) and their M+ combat log lock is released.

Reply:
Actually maybe this is just an outdated comment and the 30s should be the timeout that the constructor takes.

Reply:
In which case this is likely fine. On second read I think it is.

Reply:
The CombatLogParser class does not concert itself with what happens when the timeout reaches, except it "unlocks" the combat log such that we'll be able to receive data from any other watches logs.

This in turn fires an event (DataTimeout) that logutils catches and can act on, to stop recording for example.

The 30 seconds is a remnant - the timeout is actually set in the constructor and currently at 2 minutes. This could be made an advanced Setting so people can control it. 2 minutes might be a bit too short sometimes.